addons/survey/controllers/survey_session_manage.py

```diff
@@ -19,21 +19,31 @@ def _fetch_from_token(self, survey_token):
         return request.env['survey.survey'].search([('access_token', '=', survey_token)])
 
     def _fetch_from_session_code(self, session_code):
-        """ Matches a survey against a passed session_code.
+        """ Matches a survey against a passed session_code, and checks if it is valid.
+        If it is valid, returns the start url. Else, the error type.
         We don't limit the session_state to be reachable (ready / in_progress) here because
-        in some cases, we want closed session as well (where session_state = False).
+        in some cases, we want not yet launched and closed sessions as well (where session_state = False).
         Instead, when necessary, the reachability is forced in routes calling this method to
-        avoid people using those routes to access other (private) surveys.
-        We limit to sessions opened within the last 7 days to avoid potential abuses. """
+        avoid people using those routes to access other (private) surveys."""
         if session_code:
-            matching_survey = request.env['survey.survey'].sudo().search([
-                ('session_start_time', '>', fields.Datetime.now() - relativedelta(days=7)),
-                ('session_code', '=', session_code),
-            ], limit=1)
-            if matching_survey:
-                return matching_survey
-
-        return False
+            survey = request.env['survey.survey'].sudo().search([('session_code', '=', session_code)], limit=1)
+            if survey:
+                if survey.session_state in ['ready', 'in_progress']:
+                    return {'survey_url': '/survey/start/%s' % survey.access_token}
+                elif survey.session_start_time:
+                    # Limit to surveys closed max 7 days ago
+                    if survey.session_start_time < fields.Datetime.now() - relativedelta(days=7):
+                        return {'error': 'survey_wrong'}
+                    else:
+                        return {'error': 'survey_session_closed'}
+                elif request.env.user.has_group("survey.group_survey_user"):
+                    return {
+                        'error': 'survey_session_not_launched',
+                        'survey_id': survey.id,
+                    }
+                else:
+                    return {'error': 'survey_session_not_launched'}
```

Comment:
I would remove this block.

That way we return "survey_wrong" if the user lands on a private survey and does not have the rights to see it.
Which is the same as when we don't find anything, so we avoid leaking information.

Reply:
I'm not sure I get it.
If i remove the last block, then nobody except someone with the rights to launch the session will see the error saying that that the session is not launched yet. Isn't that going against the task spec (Detect that the code is correct and that the survey exists but that it did not start yet in case someone arrives too early) ?

Maybe should I add another verification to distinguish private from public sessions ? Or it does not matter if the error is not visible from everyone for public sessions ?

Reply:
Yes I see what you mean.

Then I would do the checks a bit differently:

```py
if survey.session_state in ['ready', 'in_progress']:
    return {'survey_url': '/survey/start/%s' % survey.access_token}
elif not survey.certification:
    survey_info = {'error': 'survey_session_not_launched'}
    if request.env.user.has_group("survey.group_survey_user"):
        survey_info['survey_id'] = survey.id  # will allow the survey user to launch the session
    return survey_info
```

We would completely remove "survey_session_closed", as it does not really matter anymore.
Indeed, it's more important to know that the session has not started yet (and allow the survey user to start it) than to know that "there were once a session for this survey but it's not active anymore".

The same survey can hold multiple live sessions, so even there was once a session some time ago, that does not mean there will not be a session starting soon.

Rest of the checks would be the same.
The reason behind the certification check is to have the same behavior as the from view, certifications cannot be started as live sessions.

We would of course also have to clean the template / JS to remove the "survey_session_closed" use case.

What do you think?

Reply:
Yeah I think it's a good approach. Could we not change the error message to handle the 2 cases, though ?
I mean we could just check if the session is live or not as proposed, and if not, display something like "session is not live at the moment" or "live session is not active" ?